docs/dev/opensearch-relevancy-search.md

```diff
@@ -0,0 +1,294 @@
+* **slop:** (integer) **** maximum number of positions allowed between matching tokens. Defaults to 0.
+
+
+The following links are redirecting to the issues pages of the search functions details including syntax, functionality and available paramters:
+
+**1. match function**
+https://github.com/opensearch-project/sql/issues/184
+
+**2. match_phrase function**
+https://github.com/opensearch-project/sql/issues/185
+
+**3. match_phrase_prefix function**
+https://github.com/opensearch-project/sql/issues/186
+
+**4. match_bool_prefix function**
+https://github.com/opensearch-project/sql/issues/187
+
+**5. multi_match function**
+https://github.com/opensearch-project/sql/issues/188
+
+**6. combined_fields function**
+https://github.com/opensearch-project/sql/issues/189
+
+**7. common function:**
+https://github.com/opensearch-project/sql/issues/190
+
+**8. query_string function**
+https://github.com/opensearch-project/sql/issues/191
+
+**9. simple_query_string function**
+https://github.com/opensearch-project/sql/issues/192
+
+
+## 5 Implementation
+The implementation covers the planner and optimizer in the query engine but skip the changes in in-memory executor as discussed in the design for current phase. One of the tricky things in the implementations is how to recognize the custom values as query parameters or options.
+
+The current solution is to pass the type check and environment resolve for the parameter values since they are not participating the in memory computing, but take the values all as string expressions and resolve them in the layer of optimizer when translated to query DSL. Take `analyzer` as an example:
+```
+// define the analyzer action
+private final BiFunction<MatchQueryBuilder, ExprValue, MatchQueryBuilder> analyzer =
+    (b, v) -> b.analyzer(v.stringValue());
+
+...
+
+ImmutableMap<Object, Object> argAction = ImmutableMap.builder()
+    .put("analyzer", analyzer)
+    .put(...)
+    .build();
+
+...
+
+// build match query
+while (iterator.hasNext()) {
+  NamedArgumentExpression arg = (NamedArgumentExpression) iterator.next();
+  if (!argAction.containsKey(arg.getArgName())) {
+    throw new SemanticCheckException(String
+        .format("Parameter %s is invalid for match function.", arg.getArgName()));
+  }
+  ((BiFunction<MatchQueryBuilder, ExprValue, MatchQueryBuilder>) argAction
+      .get(arg.getArgName()))
+      .apply(queryBuilder, arg.getValue().valueOf(null));
+}
+```
+
+## 6 Testing
+All the code changes should be test driven. The pull requests should include unit test cases, integration test cases and comparison test cases in applicable.
+
+
+
+## Appendix
+
+### A1. Relevance based search queries in OpenSearch
+
```

Comment:
There are much more parameters, check lists in [code](https://github.com/opensearch-project/sql/blob/main/opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/FunctionParameterRepository.java) and in [docs](https://opensearch.org/docs/latest/opensearch/query-dsl/full-text/#other-advanced-options)

Proposal: highlight parameter names by backticks